src/TextInputWithTokens.tsx

```diff
@@ -0,0 +1,258 @@
+import React, {FocusEventHandler, KeyboardEventHandler, useRef, useState} from 'react'
+import {omit} from '@styled-system/props'
+import styled from 'styled-components'
+import {FocusKeys} from './behaviors/focusZone'
+import {useCombinedRefs} from './hooks/useCombinedRefs'
+import {useFocusZone} from './hooks/useFocusZone'
+import {ComponentProps} from './utils/types'
+import Token from './Token/Token'
+import {TokenSizeKeys} from './Token/TokenBase'
+import TextInput, {TextInputProps} from './TextInput'
+import {useProvidedRefOrCreate} from './hooks'
+import UnstyledTextInput from './_UnstyledTextInput'
+
+const InputWrapper = styled.div`
+  order: 1;
+  flex-grow: 1;
+`
+
+// type AnyTokenProps = Partial<TokenProps & IssueLabelTokenProps & ProfileTokenProps>
+// type TokenDatum = MandateProps<AnyTokenProps, 'id' | 'text'>
+type TextInputWithTokensInternalProps<TokenComponentType extends React.ComponentType<any>> = {
+  /**
+   * The array of tokens to render
+   */
+  tokens: TokenComponentType extends React.ComponentType<infer TokenProps> ? TokenProps[] : never
+  /**
+   * The function that gets called when a token is removed
+   */
+  onTokenRemove: (tokenId: string | number) => void
+  /**
+   * The component used to render each token
+   */
+  tokenComponent?: TokenComponentType
+  /**
+   * The maximum height of the component. If the content in the input exceeds this height,
+   * it will scroll vertically
+   */
+  maxHeight?: React.CSSProperties['maxHeight']
+  /**
+   * Whether tokens should render inline horizontally. By default, tokens wrap to new lines.
+   */
+  preventTokenWrapping?: boolean
+  /**
+   * The size of the tokens
+   */
+  size?: TokenSizeKeys
+  /**
+   * Whether the remove buttons should be rendered in the tokens
+   */
+  // TODO: confirm whether or not we're comfortable with allowing people to hide that button
+  hideTokenRemoveButtons?: boolean
+} & TextInputProps
+
+// using forwardRef is important so that other components (ex. Autocomplete) can use the ref
+function TextInputWithTokensInnerComponent<TokenComponentType extends React.ComponentType<any>>(
+  {
+    icon: IconComponent,
+    contrast,
+    className,
+    block,
+    disabled,
+    theme,
+    sx: sxProp,
+    tokens,
+    onTokenRemove,
+    tokenComponent: TokenComponent,
+    preventTokenWrapping,
+    size,
+    hideTokenRemoveButtons,
+    selectedTokenIdx,
+    setSelectedTokenIdx,
+    ...rest
+  }: TextInputWithTokensInternalProps<TokenComponentType> & {
+    selectedTokenIndex: number | undefined
```

Comment:
Why is this state passed into the component instead of being managed internally?

Reply:
Because it's being used in the `useFocusZone` hook in this `TextInputWithTokensComponent`

Reply:
`TextInputWithTokensInnerComponent` renders everything inside the grey border. 
`TextInputWithTokensComponent` renders the entire thing - grey border and all.

I'm a little foggy on _why_ I had to split it like this, but I think it had to do with passing refs. I'm going to try and combine them again to remind myself why I did it this way.

If there really is a reason they need to be separate, I'll document as clearly as possible using comments.

Reply:
We may be able to merge these into one component, but I haven't figured it out quite yet.
Here's why these are separate:

1. To be able to pass a component with forwarded refs to the `as` prop of the `TextInput` component
2. To pass the ref from `useFocusZone` directly to the `TextInput`'s `Wrapper` component